ext/mbstring/mbstring.c

```diff
@@ -2982,84 +2982,81 @@ static unsigned int mb_estimate_encoding_demerits(uint32_t w)
 	return 0;
 }
 
-/* When doing 'strict' detection, any string which is invalid in the candidate encoding
- * is rejected. With non-strict detection, we just continue, but apply demerits for
- * each invalid byte sequence */
-static const mbfl_encoding* mb_guess_encoding(unsigned char *in, size_t in_len, const mbfl_encoding **elist, unsigned int elist_size, bool strict)
+struct candidate {
+	const mbfl_encoding *enc;
+	unsigned char *in;
+	size_t in_len;
+	uint64_t demerits; /* Wide bit size to prevent overflow */
+	unsigned int state;
+};
+
+static size_t init_candidate_array(struct candidate *array, size_t length, const mbfl_encoding **encodings, unsigned char **in, size_t *in_len, size_t n, bool strict)
 {
-	if (elist_size == 0) {
-		return NULL;
-	}
-	if (elist_size == 1) {
-		if (strict) {
-			return php_mb_check_encoding((const char*)in, in_len, *elist) ? *elist : NULL;
-		} else {
-			return *elist;
+	unsigned int j = 0;
+
+	for (unsigned int i = 0; i < length; i++) {
```

Comment:
```suggestion
	for (size_t i = 0; i < length; i++) {
```
Because ``length`` is a ``size_t`` I think this makes more sense as otherwise long string can technically overflow as AFAIK ``sizeof(unsingned int) != sizeof(size_t)``